connection.go

```diff
@@ -31,6 +40,13 @@ type mysqlConn struct {
 	sequence         uint8
 	parseTime        bool
 	strict           bool
+	watcher          chan<- mysqlContext
+	closech          chan struct{}
+	finished         chan<- struct{}
+
+	mu          sync.Mutex // guards following fields
+	closed      bool       // set true when conn is closed, before closech is closed
```

Comment:
Very good point, but Go currently doesn't have any atomic bool type.
I would send another PR with an implementation based e.g. on uint32 to replace it after this is merged.